qutebrowser/components/braveadblock.py

```diff
@@ -0,0 +1,325 @@
+
+class BraveAdBlocker:
+
+    """Manage blocked hosts based from /etc/hosts-like files.
+
+    Attributes:
+        _in_progress: The DownloadItems which are currently downloading.
+        _done_count: How many files have been read successfully.
+        _has_basedir: Whether a custom --basedir is set.
+        _cache_path: The path of the adblock engine cache file
+        _engine_factory: A function that can be called to create an empty
+                         adblocking engine.
+        _engine: Brave ad-blocking engine.
+    """
+
+    def __init__(
+        self,
+        *,
+        # FIXME: This type should be `adblock.Engine`. I don't know how to
+        # annotate it in such a way that mypy doesn't complain about it being
+        # undefined, since `adblock` is an optional dependency.
+        engine_factory: typing.Callable[[], typing.Any],
+        data_dir: pathlib.Path,
+        has_basedir: bool = False
+    ) -> None:
+        self._has_basedir = has_basedir
+        self._in_progress = []  # type: typing.List[downloads.TempDownload]
+        self._done_count = 0
+        self._cache_path = str(data_dir / "adblock-cache.dat")
+        self._engine_factory = engine_factory
+        self._engine = engine_factory()
+
+    def _is_blocked(
+        self,
+        request_url: QUrl,
+        first_party_url: typing.Optional[QUrl] = None,
+        resource_type: typing.Optional[interceptor.ResourceType] = None,
+    ) -> bool:
+        """Check whether the given request is blocked."""
+        if first_party_url is not None and not first_party_url.isValid():
+            first_party_url = None
+
+        if not first_party_url:
+            return False
+
+        qtutils.ensure_valid(request_url)
+
+        if not config.get("content.blocking.adblock.enabled", url=first_party_url):
+            # Do nothing if adblocking is disabled.
+            return False
+
+        result = self._engine.check_network_urls(
+            request_url.toString(),
+            first_party_url.toString(),
+            resource_type_to_string(resource_type),
+        )
+
+        if not result.matched:
+            return False
+        if result.exception is not None and result.important is None:
+            logger.debug(
+                "Excepting {} from being blocked by {} because of {}".format(
+                    request_url.toDisplayString(), result.filter, result.exception
+                )
+            )
+            return False
+        if _is_whitelisted_url(request_url):
+            logger.debug(
+                "Request to {} is whitelisted, thus not blocked".format(
+                    request_url.toDisplayString()
+                )
+            )
+            return False
+        return True
+
+    def filter_request(self, info: interceptor.Request) -> None:
+        """Block the given request if necessary."""
+        if self._is_blocked(info.request_url, info.first_party_url, info.resource_type):
+            logger.info(
```

Comment:
```suggestion
            logger.debug(
```

for consistency with #5474